File: packages/forge-canvas-dart/lib/src/canvas_registry.dart
import 'package:flutter/widgets.dart';

/// One canvas component — a name, a Flutter widget builder, and an
/// optional JSON schema for its props.
class CanvasComponent {
  final String name;
  final Widget Function(Map<String, dynamic> props) builder;
  final Map<String, dynamic>? propsSchema;

  const CanvasComponent({
    required this.name,
    required this.builder,
    this.propsSchema,
  });
}

/// Resolves component names (from backend payloads) to Flutter widget
/// builders. Mirrors the Vue/Svelte registry shape.
class CanvasRegistry {
  final Map<String, CanvasComponent> _entries = {};

  CanvasRegistry([List<CanvasComponent> initial = const []]) {
    for (final e in initial) {
      register(e);
    }
  }

  void register(CanvasComponent entry) {
    if (_entries.containsKey(entry.name)) {
      throw StateError('canvas component "${entry.name}" is already registered');
    }
    _entries[entry.name] = entry;
  }

  CanvasComponent? resolve(String name) => _entries[name];

  Iterable<CanvasComponent> entries() => _entries.values;
}

File: packages/canvas-svelte/src/canvas-registry.ts
// Canvas component registry — Svelte 5 variant.

import type { Component } from 'svelte'

export interface CanvasComponent<Props = Record<string, unknown>> {
  name: string
  component: Component
  propsSchema?: Record<string, unknown>
}

export interface CanvasRegistry {
  register(entry: CanvasComponent): void
  resolve(name: string): CanvasComponent | null
  entries(): readonly CanvasComponent[]
}

export function createCanvasRegistry(initial: CanvasComponent[] = []): CanvasRegistry {
  const entries = new Map<string, CanvasComponent>()
  for (const e of initial) entries.set(e.name, e)

  return {
    register(entry) {
      if (entries.has(entry.name)) {
        throw new Error(`canvas component "${entry.name}" is already registered`)
      }
      entries.set(entry.name, entry)
    },
    resolve(name) {
      return entries.get(name) ?? null
    },
    entries() {
      return Array.from(entries.values())
    },
  }
}
